plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceUpdate.java

```diff
@@ -0,0 +1,178 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static io.trino.plugin.iceberg.IcebergColumnHandle.ROW_ID_COLUMN_NAME;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+
+public class IcebergPageSourceUpdate
+        implements UpdatablePageSource
+{
+    private final Slice filePathSlice;
+    private final List<IcebergColumnHandle> queriedColumns;
+    private final List<IcebergColumnHandle> allTableColumns;
+    private final List<IcebergColumnHandle> updateColumns;
+    private final ConnectorPageSource source;
+    private final ConnectorPageSink posDeleteSink;
+    private final ConnectorPageSink updateRowSink;
+
+    public IcebergPageSourceUpdate(
+            String filePath,
+            List<IcebergColumnHandle> queriedColumns,
+            List<IcebergColumnHandle> allTableColumns,
+            List<IcebergColumnHandle> updateColumns,
+            ConnectorPageSource source,
+            ConnectorPageSink posDeleteSink,
+            ConnectorPageSink updateRowSink)
+    {
+        this.filePathSlice = Slices.utf8Slice(requireNonNull(filePath, "filePath is null"));
+        this.queriedColumns = requireNonNull(queriedColumns, "queriedColumns is null");
+        this.allTableColumns = requireNonNull(allTableColumns, "allTableColumns is null");
+        this.updateColumns = requireNonNull(updateColumns, "updateColumns is null");
+        this.source = requireNonNull(source, "source is null");
+        this.posDeleteSink = requireNonNull(posDeleteSink, "posDeleteSink is null");
+        this.updateRowSink = requireNonNull(updateRowSink, "updateRowSink is null");
+    }
+
+    @Override
+    public void deleteRows(Block rowIds)
+    {
+        RowBlock rows = (RowBlock) rowIds;
+        // TODO: need to use proper way to get blocks inside the row block instead of using getRawFieldBlocks(), this is just for POC purpose
+        posDeleteSink.appendPage(new Page(rows.getPositionCount(), rows.getRawFieldBlocks()));
+    }
+
+    @Override
+    public void updateRows(Page page, List<Integer> columnValueAndRowIdChannels)
+    {
+        RowBlock rowIdBlock = (RowBlock) page.getBlock(columnValueAndRowIdChannels.get(columnValueAndRowIdChannels.size() - 1));
+        deleteRows(rowIdBlock);
+        Map<Integer, Block> updatedColumnBlocks = new HashMap<>();
+        for (int i = 0; i < columnValueAndRowIdChannels.size() - 1; i++) {
+            IcebergColumnHandle updatedColumn = updateColumns.get(i);
+            Block updatedValues = page.getBlock(columnValueAndRowIdChannels.get(i));
+            updatedColumnBlocks.put(updatedColumn.getId(), updatedValues);
+        }
+
+        Block[] updatedRows = new Block[allTableColumns.size()];
+        Block[] oldRows = ((RowBlock) rowIdBlock.getRawFieldBlocks()[2]).getRawFieldBlocks();
```

Comment:
Cast to `RowBlock` isn't entirely correct.
See #9354 and perhaps we should use `ColumnarRow` here. 
cc @djsstarburst